handler/OpenIDHandler.php

```diff
@@ -0,0 +1,464 @@
+     * This is a very simple step, and it should be safe because the token is valid at this point.
+     * If the token is invalid, the auth process stops before this function is called.
+     *
+     * @param array $credentials
+     * @return User|null
+     */
+    private function _getUserViaKeycloakId(array $credentials)
+    {
+        $userIds = Repo::user()->getIdsBySetting('openid::'.$credentials['selectedProvider'], $credentials['id']);
+        if ($userIds->isNotEmpty()) {
+            return Repo::user()->get($userIds->firstOrFail());
+        }
+        // prior versions of this plugin used hash for saving the openid identifier, but this is not recommended.
+        $userIds = Repo::user()->getIdsBySetting('openid::'.$credentials['selectedProvider'], hash('sha256', $credentials['id']));
+        if ($userIds->isNotEmpty()) {
+            return Repo::user()->get($userIds->firstOrFail()->getId());
+        }
+
+        return null;
+    }
+
+
+    /**
+     * This function swaps the Auth code into a JWT that contains the user_details and a signature.
+     * An array with the access_token, id_token and/or refresh_token is returned on success, otherwise null.
+     * The OpenID implementation differs a bit between the providers. Some use an id_token, others a refresh token.
+     *
+     * @param array $providerList
+     * @param string $authorizationCode
+     * @param string $selectedProvider
+     * @return array
+     */
+    private function _getTokenViaAuthCode(array $providerList, string $authorizationCode, string $selectedProvider)
+    {
+        $token = null;
+        if (isset($providerList) && key_exists($selectedProvider, $providerList)) {
+            $settings = $providerList[$selectedProvider];
+            $httpClient = Application::get()->getHttpClient();
+            $response = null;
+            $params = [
+                'code' => $authorizationCode,
+                'grant_type' => 'authorization_code',
+                'client_id' => $settings['clientId'],
+                'client_secret' => $settings['clientSecret'],
+            ];
+            if ($selectedProvider != 'microsoft') {
+                $params['redirect_uri'] = Application::get()->getRequest()->url(
+                    null,
+                    'openid',
+                    'doAuthentication',
+                    null,
+                    array('provider' => $selectedProvider)
+                );
+            }
+            try {
+                $response = $httpClient->request(
+                    'POST',
+                    $settings['tokenUrl'],
+                    [
+                        'headers' => [
+                            'Accept' => 'application/json',
+                        ],
+                        'form_params' => $params,
+                    ]
+                );
+                if ($response->getStatusCode() != 200) {
+                    error_log('Guzzle Response != 200: '.$response->getStatusCode());
```

Comment:
Add a prefix to error_log calls, just to hint that it's coming from this plugin, since they won't include a stack trace.

Reply:
Added this `error_log($this->plugin->getName() . ' - Guzzle Response != 200: ' . $response->getStatusCode());` here and elsewhere - maybe also the exact call that fails could be in order here